File: Arrays/240-Search-a-2D-Matrix-II/JavaScript/240-Search-a-2D-Matrix-II-Optimal.js
/**
 * Problem: 240. Search a 2D Matrix II
 * Approach: Optimal - Staircase Search
 *
 * Start from the top-right corner.
 *
 * If current element is smaller than target, move down.
 * If current element is larger than target, move left.
 *
 * Time Complexity: O(m + n)
 * Space Complexity: O(1)
 */

/**
 * @param {number[][]} matrix
 * @param {number} target
 * @return {boolean}
 */
var searchMatrix = function(matrix, target) {
    const m = matrix.length;
    const n = matrix[0].length;

    let row = 0;
    let col = n - 1;

    while (row < m && col >= 0) {
        if (matrix[row][col] === target) {
            return true;
        } else if (matrix[row][col] < target) {
            row++;
        } else {
            col--;
        }
    }

    return false;
};

File: Arrays/240-Search-a-2D-Matrix-II/Python/240-Search-a-2D-Matrix-II-Optimal.py
"""
Problem: 240. Search a 2D Matrix II
Approach: Optimal - Staircase Search

Start from the top-right corner.

If current element is smaller than target, move down.
If current element is larger than target, move left.

Each step eliminates one row or one column.

Time Complexity: O(m + n)
Space Complexity: O(1)
"""

class Solution:
    def searchMatrix(self, matrix: list[list[int]], target: int) -> bool:
        m = len(matrix)
        n = len(matrix[0])

        row = 0
        col = n - 1

        while row < m and col >= 0:
            if matrix[row][col] == target:
                return True
            elif matrix[row][col] < target:
                row += 1
            else:
                col -= 1

        return False